### ISSUE TITLE
`MockAgent`: `MockResponseCallbackOptions` given to `replyOptionsCallback` is a totally wrong type

### ISSUE THREAD

**Issue by jonahsnider** [nodejs/undici#1583]
## Reproducible By

```ts
import * as undici from 'undici';

const mockAgent = new undici.MockAgent();
mockAgent.disableNetConnect();
undici.setGlobalDispatcher(mockAgent);

const mockPool = mockAgent.get('https://example.com');

mockPool
  .intercept({
    path: '/',
  })
  .reply((request) => {
    console.log(request);

    return { statusCode: 200 };
  });

await undici.request('https://example.com/', {
  headers: {
    authorization: 'Bearer 123',
  },
});
```

## Expected Behavior

The types should be correct.

## Logs & Screenshots

Actual value:

```js
{
  headers: { authorization: 'Bearer 123' },
  origin: 'https://example.com',
  path: '/',
  method: 'GET'
}
```

Expected value:

https://github.com/nodejs/undici/blob/1822ee6eea459bb7432996064144c394d27ee03f/types/mock-interceptor.d.ts#L26

https://github.com/nodejs/undici/blob/1822ee6eea459bb7432996064144c394d27ee03f/types/mock-interceptor.d.ts#L85-L87

https://github.com/nodejs/undici/blob/1822ee6eea459bb7432996064144c394d27ee03f/types/mock-interceptor.d.ts#L72-L79

## Environment

Darwin 21.5.0 Darwin Kernel Version 21.5.0: Tue Apr 26 21:08:37 PDT 2022; root:xnu-8020.121.3~4/RELEASE_ARM64_T6000 arm64

```js
{
  node: '16.15.1',
  v8: '9.4.146.24-node.21',
  uv: '1.43.0',
  zlib: '1.2.11',
  brotli: '1.0.9',
  ares: '1.18.1',
  modules: '93',
  nghttp2: '1.47.0',
  napi: '8',
  llhttp: '6.0.4',
  openssl: '1.1.1o+quic',
  cldr: '40.0',
  icu: '70.1',
  tz: '2021a3',
  unicode: '14.0',
  ngtcp2: '0.1.0-DEV',
  nghttp3: '0.1.0-DEV'
}
```

**Comment by mcollina** [comment]
Thanks for reporting! Would you like to send a Pull Request to address this issue? Remember to add unit tests.

**Comment by sidwebworks** [comment]
@mcollina if you want I can take this issue?

**Comment by mcollina** [comment]
go for it!

**Comment by merojosa** [comment]
@mcollina, I'm currently trying to fix this issue. I reproduced the code and, indeed,`reply` is receiving the following:

```
{
  headers: { authorization: 'Bearer 123' },
  origin: 'https://example.com',
  path: '/',
  method: 'GET'
}
```

However, the type for the callback says that it can have these options:

```
export interface MockResponseCallbackOptions {
  path: string;
  origin: string;
  method: string;
  body?: BodyInit | Dispatcher.DispatchOptions['body'];
  headers: Headers | Record<string, string>;
  maxRedirections: number;
}
```

`maxRedirections` and `body` are not been returned. I already checked the [documentation](https://undici.nodejs.org/#/docs/api/MockAgent) and don't see any particular example or explanation regarding if it's posible to get the `body` and `maxRedirections`.

So I want to check if this is a mismatch between the type and what `reply` can receive in its parameters or there is a bug with `MockAgent`.
